core/arch/riscv64/emit_utils.c

```diff
@@ -587,14 +587,255 @@ emit_inline_ibl_stub(dcontext_t *dcontext, byte *pc, ibl_code_t *ibl_code,
+
+    /* The hash "function":
+     * a1: hash mask
+     * a2: indirect branch target
+     */
+    APP(&ilist,
+        INSTR_CREATE_and(dc, opnd_create_reg(DR_REG_A1), opnd_create_reg(DR_REG_A1),
+                         opnd_create_reg(DR_REG_A2)));
+
+    /* Now, a1 holds the hash table index, use slli+add to get the table entry. */
+    ASSERT(4 - HASHTABLE_IBL_OFFSET(ibl_code->branch_type) >= 0);
+    if (4 - HASHTABLE_IBL_OFFSET(ibl_code->branch_type) > 0) {
+        APP(&ilist,
+            INSTR_CREATE_slli(
+                dc, opnd_create_reg(DR_REG_A1), opnd_create_reg(DR_REG_A1),
+                opnd_add_flags(
+                    opnd_create_immed_int(4 - HASHTABLE_IBL_OFFSET(ibl_code->branch_type),
+                                          OPSZ_6b),
+                    DR_OPND_IMM_PRINT_DECIMAL)));
+    }
+    APP(&ilist,
+        INSTR_CREATE_add(dc, opnd_create_reg(DR_REG_A1), opnd_create_reg(DR_REG_A0),
+                         opnd_create_reg(DR_REG_A1)));
+
+    /* Jump back from sentinel when wraparound is needed. */
+    APP(&ilist, load_tag);
+
+    /* a1: table entry (fragment_entry_t*).
+     * Load tag_fragment from fragment_entry_t* into a0.
+     */
+    APP(&ilist,
+        INSTR_CREATE_ld(
+            dc, opnd_create_reg(DR_REG_A0),
+            OPND_CREATE_MEMPTR(DR_REG_A1, offsetof(fragment_entry_t, tag_fragment))));
+
+    /* Jump back from collision. */
+    APP(&ilist, compare_tag);
+
+    /* a0: tag_fragment
+     * a2: indirect branch target
+     * Did we hit?
+     */
+    APP(&ilist,
+        INSTR_CREATE_beq(dc, opnd_create_instr(not_hit), opnd_create_reg(DR_REG_A0),
+                         opnd_create_reg(DR_REG_X0)));
+    /* We hit, but did it collide? */
+    APP(&ilist,
+        INSTR_CREATE_bne(dc, opnd_create_instr(try_next), opnd_create_reg(DR_REG_A0),
+                         opnd_create_reg(DR_REG_A2)));
+
+    /* No, so we found the answer.
+     * App's original values of a0 and a1 are already in respective TLS slots, and
+     * will be restored by the fragment prefix.
+     */
+
+    /* Recover app's original a2. */
+    APP(&ilist, instr_create_restore_from_tls(dc, DR_REG_A2, TLS_REG2_SLOT));
+
+    /* Load the answer into a0 */
+    APP(&ilist,
+        INSTR_CREATE_ld(dc, opnd_create_reg(DR_REG_A0),
+                        OPND_CREATE_MEMPTR(
+                            DR_REG_A1, offsetof(fragment_entry_t, start_pc_fragment))));
+    /* jr a0
+     * (keep in sync with instr_is_ibl_hit_jump())
+     */
+    APP(&ilist, XINST_CREATE_jump_reg(dc, opnd_create_reg(DR_REG_A0)));
+
+    APP(&ilist, try_next);
+
+    /* Try next entry, in case of collision. No wraparound check is needed
+     * because of the sentinel at the end.
+     */
+    APP(&ilist,
+        INSTR_CREATE_addi(dc, opnd_create_reg(DR_REG_A1), opnd_create_reg(DR_REG_A1),
+                          opnd_create_immed_int(sizeof(fragment_entry_t), OPSZ_12b)));
```

Comment:
Could we have these immediate sizes auto-figured-out?  This seems error-prone for a user to know that this should be a 12-bit immed.  (It is true that other arches like x86 do have the user deciding between 1-byte or 4-byte immeds in some cases IIRC: but that seems like a different level of decision (and we would like to auto-magic that away too) than the precise encoding field bit width.)

Reply:
Ditto.

Reply:
OK, are these issues written down somewhere, maybe in the #3544 container case?  Just want to not lose the action item.

Reply:
Could we also have a TODO i#3544 comment here so the reader knows that we realize this is not ideal and not what we want long-term.